File: learning-factory/warehouse/artifacts/projects/build-your-own-x/docker/a-proof-of-concept-imitation-of-docker-written-in-100-python-482ca0b333/job_byox_build_v1_28dac8305a64d1c464faa8e07190fb53/attempt-001/sealed/reference_tests/validate_pack.py
"""Reproducible structural and leak checks for the generated challenge pack."""

from __future__ import annotations

import ast
import hashlib
import json
import os
from pathlib import Path
import re
import stat


REQUIRED_PATHS = (
    "README.md",
    "AGENTS.md",
    "MANIFEST.yaml",
    "PROVENANCE.json",
    "LICENSE_BOUNDARY.md",
    "REQUIREMENTS.md",
    "CONCEPTS.md",
    "DESIGN_QUESTIONS.md",
    "VALIDATION.md",
    "starter/README.md",
    "public_tests/README.md",
    "environment/README.md",
    "sealed/reference/README.md",
    "sealed/reference_tests/README.md",
    "sealed/DESIGN.md",
    "sealed/TRADEOFFS.md",
    "sealed/REVIEW.md",
    "sealed/alternatives/README.md",
    "sealed/production/PRODUCTIONIZATION.md",
    "adversarial/README.md",
    "debugging/README.md",
    "review_exercises/README.md",
    "benchmarks/README.md",
)

FORBIDDEN_PATHS = (
    ".git",
    ".env",
    ".venv",
    "credentials.json",
    "secrets",
    "reference",
    "reference_tests",
    "hidden_tests",
    "solution",
    "solutions",
    "answers",
    "starter/sealed",
    "starter/reference",
    "starter/reference_tests",
    "starter/solution",
    "starter/solutions",
    "starter/answers",
    "public_tests/sealed",
    "public_tests/reference",
    "public_tests/hidden_tests",
    "environment/sealed",
)

GENERATED_ROOTS = (
    "starter",
    "public_tests",
    "environment",
    "sealed",
    "adversarial",
    "debugging",
    "review_exercises",
    "benchmarks",
)

MANIFEST = {
    "independent_validation": "REQUIRED",
    "productionized": False,
    "project_id": "project_d76f888c329de3f5823edf9ffcbe85c3",
    "provenance_sha256": "b87d00ac1851cdd19fddde57f22c054a6b66257b6bdce4c4cbaf3c2bce3516c3",
    "schema_version": 1,
    "source_commit": "aa17439b62f384511a5561ce308e9598b94d8989",
    "source_id": "source_eac489a34bed5db9a1f2a580b457bcef",
    "status": "GENERATED",
    "validation_labels": ["GENERATED", "PARTIAL"],
}


def strict_json(path: Path) -> object:
    def unique_object(pairs: list[tuple[str, object]]) -> dict[str, object]:
        result: dict[str, object] = {}
        for key, value in pairs:
            if key in result:
                raise AssertionError(f"duplicate JSON key {key!r} in {path}")
            result[key] = value
        return result

    return json.loads(path.read_text(encoding="utf-8"), object_pairs_hook=unique_object)


def main() -> int:
    root = Path.cwd()
    missing = [name for name in REQUIRED_PATHS if not (root / name).is_file()]
    forbidden = [name for name in FORBIDDEN_PATHS if os.path.lexists(root / name)]
    if missing or forbidden:
        raise AssertionError({"missing": missing, "forbidden": forbidden})

    manifest = strict_json(root / "MANIFEST.yaml")
    if manifest != MANIFEST:
        raise AssertionError("MANIFEST.yaml differs from the immutable expected object")
    provenance = strict_json(root / "PROVENANCE.json")
    if not isinstance(provenance, dict):
        raise AssertionError("PROVENANCE.json must contain an object")
    if set(provenance) != {
        "classification",
        "license_boundary",
        "project",
        "schema_version",
        "snapshot_sha256",
        "source",
    }:
        raise AssertionError("unexpected top-level provenance fields")
    if provenance["snapshot_sha256"] != MANIFEST["provenance_sha256"]:
        raise AssertionError("manifest/provenance snapshot mismatch")
    if provenance["project"]["project_id"] != MANIFEST["project_id"]:  # type: ignore[index]
        raise AssertionError("manifest/provenance project mismatch")
    if provenance["license_boundary"]["linked_content_copied"] is not False:  # type: ignore[index]
        raise AssertionError("license boundary does not deny linked-content copying")

    all_paths: list[Path] = []
    for name in GENERATED_ROOTS:
        generated_root = root / name
        all_paths.extend((generated_root, *generated_root.rglob("*")))
    symlinks: list[str] = []
    special: list[str] = []
    for path in all_paths:
        mode = path.lstat().st_mode
        relative = str(path.relative_to(root))
        if stat.S_ISLNK(mode):
            symlinks.append(relative)
        elif not (stat.S_ISREG(mode) or stat.S_ISDIR(mode)):
            special.append(relative)
    if symlinks or special:
        raise AssertionError({"symlinks": symlinks, "special": special})

    answer_paths = [path for path in all_paths if path.is_file() and path.name == "ANSWER.md"]
    if any(path.relative_to(root).parts[0] != "sealed" for path in answer_paths):
        raise AssertionError("an exercise answer exists outside the sealed tree")

    credential_patterns = (
        re.compile(r"-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----"),
        re.compile(r"\b(?:AKIA|ASIA)[A-Z0-9]{16}\b"),
        re.compile(r"\bgh[pousr]_[A-Za-z0-9]{30,}\b"),
        re.compile(r"\bxox[baprs]-[A-Za-z0-9-]{20,}\b"),
        re.compile(
            r'''(?i)\b(?:api[_-]?key|access[_-]?token|password)\s*[:=]\s*['"][^'"]{8,}['"]'''
        ),
    )
    text_files = [path for path in all_paths if path.is_file()]
    text_files.extend(root / name for name in REQUIRED_PATHS if (root / name).parent == root)
    credential_hits: list[str] = []
    for path in set(text_files):
        content = path.read_text(encoding="utf-8")
        if any(pattern.search(content) for pattern in credential_patterns):
            credential_hits.append(str(path.relative_to(root)))
    if credential_hits:
        raise AssertionError({"possible_credentials": sorted(credential_hits)})

    policy_violations: list[str] = []
    python_files = sorted(path for path in text_files if path.suffix == ".py")
    for path in python_files:
        tree = ast.parse(path.read_text(encoding="utf-8"), filename=str(path))
        for node in ast.walk(tree):
            if not isinstance(node, ast.Call):
                continue
            call_name = (
                node.func.attr
                if isinstance(node.func, ast.Attribute)
                else node.func.id if isinstance(node.func, ast.Name) else ""
            )
            if call_name in {"extract", "extractall"}:
                policy_violations.append(f"{path}:{node.lineno}: unsafe tar helper")
            if any(
                keyword.arg == "shell"
                and isinstance(keyword.value, ast.Constant)
                and keyword.value.value is True
                for keyword in node.keywords
            ):
                policy_violations.append(f"{path}:{node.lineno}: shell=True")
    if policy_violations:
        raise AssertionError(policy_violations)

    report = {
        "answer_files_outside_sealed": 0,
        "forbidden_present": 0,
        "high_confidence_credential_hits": 0,
        "manifest_exact": True,
        "policy_violations": 0,
        "provenance_json_valid": True,
        "provenance_raw_sha256": hashlib.sha256((root / "PROVENANCE.json").read_bytes()).hexdigest(),
        "python_files_parsed": len(set(python_files)),
        "regular_only": True,
        "required_files": len(REQUIRED_PATHS),
        "symlinks": 0,
    }
    print(json.dumps(report, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/web-server/building-a-simple-websockets-server-from-scratch-in-ruby-c5db2d536e/job_byox_build_v1_92c6d6b7201d4d460ce04f69dcffd19c/attempt-001/sealed/validate_pack.rb
# frozen_string_literal: true

require "digest"
require "json"

ROOT = File.expand_path("..", __dir__)

REQUIRED = %w[
  README.md
  AGENTS.md
  MANIFEST.yaml
  PROVENANCE.json
  LICENSE_BOUNDARY.md
  REQUIREMENTS.md
  CONCEPTS.md
  DESIGN_QUESTIONS.md
  VALIDATION.md
  starter/README.md
  public_tests/README.md
  environment/README.md
  sealed/reference/README.md
  sealed/reference_tests/README.md
  sealed/DESIGN.md
  sealed/TRADEOFFS.md
  sealed/REVIEW.md
  sealed/alternatives/README.md
  sealed/production/PRODUCTIONIZATION.md
  adversarial/README.md
  debugging/README.md
  review_exercises/README.md
  benchmarks/README.md
].freeze

FORBIDDEN = %w[
  .git
  .env
  .venv
  credentials.json
  secrets
  reference
  reference_tests
  hidden_tests
  solution
  solutions
  answers
  starter/sealed
  starter/reference
  starter/reference_tests
  starter/solution
  starter/solutions
  starter/answers
  public_tests/sealed
  public_tests/reference
  public_tests/hidden_tests
  environment/sealed
].freeze

EXPECTED_MANIFEST = {
  "independent_validation" => "REQUIRED",
  "productionized" => false,
  "project_id" => "project_0481ac5d3348b421d2f9d5d0c6d52ecb",
  "provenance_sha256" => "aa3c412b3df6335355c1c77e7d9226a99d11df1604e54a4369e515acd2b99773",
  "schema_version" => 1,
  "source_commit" => "aa17439b62f384511a5561ce308e9598b94d8989",
  "source_id" => "source_eac489a34bed5db9a1f2a580b457bcef",
  "status" => "GENERATED",
  "validation_labels" => ["GENERATED", "PARTIAL"]
}.freeze

failures = []

REQUIRED.each do |relative|
  path = File.join(ROOT, relative)
  failures << "missing regular file: #{relative}" unless File.file?(path) && !File.symlink?(path)
end

FORBIDDEN.each do |relative|
  path = File.join(ROOT, relative)
  failures << "forbidden path exists: #{relative}" if File.exist?(path) || File.symlink?(path)
end

Dir.glob(File.join(ROOT, "**", "*"), File::FNM_DOTMATCH).each do |path|
  next if [".", ".."].include?(File.basename(path))
  stat = File.lstat(path)
  relative = path.sub(%r{\A#{Regexp.escape(ROOT)}/?}, "")
  failures << "symlink is not archivable: #{relative}" if stat.symlink?
  unless stat.file? || stat.directory?
    failures << "special file is not archivable: #{relative}"
  end
end

begin
  manifest = JSON.parse(File.binread(File.join(ROOT, "MANIFEST.yaml")))
  failures << "MANIFEST.yaml object differs" unless manifest == EXPECTED_MANIFEST
rescue JSON::ParserError => error
  failures << "MANIFEST.yaml is not strict JSON: #{error.message}"
end

begin
  provenance = JSON.parse(File.binread(File.join(ROOT, "PROVENANCE.json")))
  linked = provenance.dig("license_boundary", "linked_content_copied")
  snapshot = provenance["snapshot_sha256"]
  project_id = provenance.dig("project", "project_id")
  failures << "provenance copied-content flag differs" unless linked == false
  failures << "provenance snapshot differs" unless snapshot == EXPECTED_MANIFEST["provenance_sha256"]
  failures << "provenance project differs" unless project_id == EXPECTED_MANIFEST["project_id"]
rescue JSON::ParserError => error
  failures << "PROVENANCE.json is not strict JSON: #{error.message}"
end

secret_patterns = {
  "PEM private key" => /-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----/,
  "AWS access key" => /AKIA[0-9A-Z]{16}/,
  "GitHub token" => /gh[pousr]_[A-Za-z0-9]{20,}/,
  "Slack token" => /xox[baprs]-[A-Za-z0-9-]{10,}/,
  "assigned credential" => /(?:password|secret|api[_-]?key|access[_-]?token)\s*[:=]\s*["'][^"'\r\n]+["']/i
}.freeze

Dir.glob(File.join(ROOT, "**", "*"), File::FNM_DOTMATCH).each do |path|
  next unless File.file?(path)
  content = File.binread(path)
  secret_patterns.each do |label, pattern|
    failures << "possible #{label}: #{path.sub(ROOT + '/', '')}" if content.match?(pattern)
  end
end

if failures.empty?
  puts "PASS structure, forbidden paths, metadata status, regular-file policy, and credential scan"
else
  failures.each { |failure| warn "FAIL #{failure}" }
  exit 1
end

